File: Behavioral/Lab14/Typescript/src/Entities/realtor.ts
import { Customer } from './customer';
import { Person } from './person';
import { RealEstate } from './real_estate';

export class RealtorAbstract extends Person {
    protected _realtor: RealtorAbstract;

    public yearsOfWork: number;
    constructor(nameAndSurname: string, phone: string, email: string, yearsOfWork: number){
        super(nameAndSurname, phone, email);
        this.yearsOfWork = yearsOfWork;
    }
    public static newRealtorFromObject(realtor: RealtorAbstract): RealtorAbstract {
        let r = new RealtorAbstract(realtor.nameAndSurname, realtor.phone, realtor.email, realtor.yearsOfWork);
        return r;
    }

    public SetNext(realtor: RealtorAbstract): RealtorAbstract {
        this._realtor = realtor;
        return this._realtor;
    }

    public SellEstate(customer: Customer, estate: RealEstate): void {
        console.log(`${this.nameAndSurname} is selling ${estate.constructor.name} to ${customer.nameAndSurname} with price ${estate.priceInUSD}$`);
    }
    public toString(): string {
        return `Realtor:\nId:${this.id}, Name and surname: ${this.nameAndSurname},\nPhone number: ${this.phone}, Email: ${this.email}, Years of work ${this.yearsOfWork}`;
    }
}

File: Structural/Lab9/C#/Entities/Realtor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace C_.Entities
{
    public class Realtor : Person
    {
        double YearsOfWork{ get; set; }
        public Realtor(string nameAndSurname, string phone, string email, double yearsOfWork):base(nameAndSurname, phone, email) {
            YearsOfWork = yearsOfWork;
        }

        public Realtor(Realtor realtor):base(realtor.NameAndSurname, realtor.Phone, realtor.Email) {
            this.Id =realtor.Id;
        }

        public override string ToString()
        {
            return $"Realtor:\nId:{Id}, Name and surname: {NameAndSurname},\nPhone number: {Phone}, Email: {Email}, Years of work {YearsOfWork}";
        }
    }
}
